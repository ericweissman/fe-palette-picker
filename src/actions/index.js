export const isLoading = (bool) => ({
  type: 'IS_LOADING',
  isLoading: bool
});

export const hasErrored = (message) => ({
  type: 'HAS_ERRORED',
  message
});

//projects
export const getProjectsSuccess = (projects) => ({
  type: 'GET_PROJECTS_SUCCESS',
  projects
})

export const addProjectSuccess = (project) => ({
  type: 'ADD_PROJECT_SUCCESS',
  project
})