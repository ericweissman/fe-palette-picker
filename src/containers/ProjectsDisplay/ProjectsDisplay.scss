.projects-display-grid {
  display: flex;
  margin-left: 15px;
  width: auto;
}

.projects-display-grid_column {
  padding-left: 30px;
  background-clip: padding-box;
}
 
.projects-display-grid_column > div {
  background: grey;
  margin-bottom: 30px;
}

.no-projects {
  color: #e6e8ed;
  font-size: 1.4rem;
  text-align: center;
  margin: 50px 0;
}