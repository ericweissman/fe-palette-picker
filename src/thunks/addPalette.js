import { isLoading, hasErrored, addPaletteSuccess } from '../actions'

export const addPalette = (url, palette) => {
  return async (dispatch) => {
    try {
      dispatch(isLoading(true))
      const response = await fetch(url, {
        method: "POST",
        headers: {
          "Content-Type": "application/json"
        },
        body: JSON.stringify(palette)
      })
      if (!response.ok) {
        throw Error(response.statusText)
      }
      dispatch(isLoading(false))
      const result = await response.json()
      dispatch(addPaletteSuccess(result))
    } catch (error) {
      dispatch(hasErrored(error.message))
    }
  }
}