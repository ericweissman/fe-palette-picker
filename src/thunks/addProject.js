import { isLoading, hasErrored, addProjectSuccess } from '../actions'

export const addProject = (url, project_name) => {
  return async (dispatch) => {
    try {
      dispatch(isLoading(true))
      const response = await fetch(url, {
        method: "POST",
        headers: {
          "Content-Type" : "application/json"
        },
        body: JSON.stringify({ project_name })
      })
      if (!response.ok) {
        throw Error(response.statusText)
      }
      dispatch(isLoading(false))
      const result = await response.json()
      dispatch(addProjectSuccess(result))
    } catch (error) {
      dispatch(hasErrored(error.message))
    }
  }
}