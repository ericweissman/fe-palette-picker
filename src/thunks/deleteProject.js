import { isLoading, hasErrored, deleteProjectSuccess } from '../actions'

export const deleteProject = (url, projectID) => {
  return async (dispatch) => {
    try {
      dispatch(isLoading(true))
      const response = await fetch(url, {
        method: "DELETE",
        headers: {
          "Content-Type" : "application/json"
        }
      })
      if (!response.ok) {
        throw Error(response.statusText)
      }
      dispatch(isLoading(false))
      // const result = await response.json()
      dispatch(deleteProjectSuccess(projectID))
    } catch (error) {
      dispatch(hasErrored(error.message))
    }
  }
}