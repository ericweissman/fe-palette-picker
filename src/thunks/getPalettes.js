import { isLoading, hasErrored, getPalettesSuccess } from '../actions'

export const getPalettes = (url, id) => {
  return async (dispatch) => {
    try {
      dispatch(isLoading(true))
      const response = await fetch(url)
      if (!response.ok) {
        throw Error(response.statusText)
      }
      dispatch(isLoading(false))
      const result = await response.json()
      dispatch(getPalettesSuccess(result, id))
    } catch (error) {
      dispatch(hasErrored(error.message))
    }
  }
}