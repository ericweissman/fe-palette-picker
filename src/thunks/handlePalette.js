import { isLoading, hasErrored } from '../actions'

export const handlePalette = (url, actionToDispatch, method, palette) => {
  return async (dispatch) => {
    try {
      dispatch(isLoading(true))
      const response = await fetch(url, {
        method: `${method}`,
        headers: {
          "Content-Type": "application/json"
        },
        body: JSON.stringify(palette)
      })
      if (!response.ok) {
        throw Error(response.statusText)
      }
      dispatch(isLoading(false))
      const result = await response.json()
      dispatch(actionToDispatch(result))
    } catch (error) {
      dispatch(hasErrored(error.message))
    }
  }
}