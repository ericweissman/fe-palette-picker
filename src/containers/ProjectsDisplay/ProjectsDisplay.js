import React, { Component } from 'react'
import { connect } from 'react-redux'
import { handlePalette } from '../../thunks/handlePalette';
import ProjectCard from '../ProjectCard/ProjectCard'

export class ProjectsDisplay extends Component {

  render() {
    const { projects } = this.props
    
    return (
      <div className="projects-container">
        {
          projects && projects.map(project => {
            return <ProjectCard key={project.id} project={project} />
          })
        }
      </div>
    )
  }
}

export const mapStateToProps = (state) => ({
  projects: state.projects,
  palettes: state.palettes,
});

export const mapDispatchToProps = (dispatch) => ({
  handlePalette: (url, actionToDispatch, method, palette) => dispatch(handlePalette(url, actionToDispatch, method, palette)),
})

export default connect(mapStateToProps, mapDispatchToProps)(ProjectsDisplay);