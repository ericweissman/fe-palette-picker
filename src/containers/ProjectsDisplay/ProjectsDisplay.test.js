import React from 'react';
import { shallow } from 'enzyme';
import { handlePalette } from '../../thunks/handlePalette'
import { ProjectsDisplay, mapStateToProps, mapDispatchToProps } from './ProjectsDisplay'
import { addPaletteSuccess } from '../../actions'

jest.mock('../../thunks/handlePalette')

describe('ProjectsDisplay', () => {
  let wrapper

  describe('snapshots', () => {

    it('should match the initial snapshot', () => {
      const mockProjects = []
      const mockPalettes = []

      wrapper = shallow(
        <ProjectsDisplay projects={mockProjects} palettes={mockPalettes} />
      )

      expect(wrapper).toMatchSnapshot()
    })

    it('should match the snapshot with projects and palettes', () => {
      const mockProjects = [{ id: 1, project_name: 'first' }]
      const mockPalettes = [{ id: 4, palette_name: 'palette 1', color_1: '111111', color_2: '222222', color_3: '333333', color_4: '444444', color_5: '555555' }]

      wrapper = shallow(
        <ProjectsDisplay projects={mockProjects} palettes={mockPalettes} />
      )

      expect(wrapper).toMatchSnapshot()
    })
  })

  describe('mapStateToProps', () => {
    it('should return an object with a projects array and palettes array', () => {
      const mockState = { projects: [{}, {}], palettes: [{}, {}, {}, {}, {}], favorites: [] }
      const expected = { projects: [{}, {}], palettes: [{}, {}, {}, {}, {}] }

      const mappedProps = mapStateToProps(mockState)

      expect(mappedProps).toEqual(expected)
    })
  })

  describe('mapDispatchToProps', () => {
    it('should call dispatch', () => {
      const mockDispatch = jest.fn()
      const mockURL = 'http://localhost:3001/api/v1/projects/1/palettes'
      const mockPalette = {}

      const actionToDispatch = handlePalette(mockURL, addPaletteSuccess, 'POST', mockPalette)

      const mappedProps = mapDispatchToProps(mockDispatch)
      mappedProps.handlePalette(mockURL, addPaletteSuccess, 'POST', mockPalette)

      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch)
    })
  })
})