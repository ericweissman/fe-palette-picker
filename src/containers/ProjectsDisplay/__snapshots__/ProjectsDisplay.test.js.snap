// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProjectsDisplay snapshots should match the initial snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <ProjectsDisplay
    palettes={Array []}
    projects={Array []}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [],
      "className": "projects-container",
    },
    "ref": null,
    "rendered": Array [],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [],
        "className": "projects-container",
      },
      "ref": null,
      "rendered": Array [],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
  Symbol(enzyme.__childContext__): null,
}
`;

exports[`ProjectsDisplay snapshots should match the snapshot with projects and palettes 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <ProjectsDisplay
    palettes={
      Array [
        Object {
          "color_1": "111111",
          "color_2": "222222",
          "color_3": "333333",
          "color_4": "444444",
          "color_5": "555555",
          "id": 4,
          "palette_name": "palette 1",
        },
      ]
    }
    projects={
      Array [
        Object {
          "id": 1,
          "project_name": "first",
        },
      ]
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <Connect(ProjectCard)
          project={
            Object {
              "id": 1,
              "project_name": "first",
            }
          }
        />,
      ],
      "className": "projects-container",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": "1",
        "nodeType": "class",
        "props": Object {
          "project": Object {
            "id": 1,
            "project_name": "first",
          },
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <Connect(ProjectCard)
            project={
              Object {
                "id": 1,
                "project_name": "first",
              }
            }
          />,
        ],
        "className": "projects-container",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": "1",
          "nodeType": "class",
          "props": Object {
            "project": Object {
              "id": 1,
              "project_name": "first",
            },
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
  Symbol(enzyme.__childContext__): null,
}
`;
